File: app/src/main/java/chenmc/sms/transaction/SmsReceiveService.java
package chenmc.sms.transaction;

import android.app.Service;
import android.content.Intent;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.IBinder;
import android.os.Looper;
import android.os.Message;
import android.os.Process;

import chenmc.sms.transaction.handler.SmsHandler;

/**
 * @author 明 明
 *         Created on 2017-4-28.
 */

public class SmsReceiveService extends Service {

    public Handler mServiceHandler;
    
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        // 新开一个线程进行处理
        HandlerThread handlerThread = new HandlerThread(getClass().getSimpleName(),
            Process.THREAD_PRIORITY_BACKGROUND);
        handlerThread.start();
    
        mServiceHandler = new ServiceHandler(handlerThread.getLooper());
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Message msg = mServiceHandler.obtainMessage();
        msg.arg1 = startId;
        msg.obj = intent;
        mServiceHandler.sendMessage(msg);

        return START_NOT_STICKY;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        mServiceHandler.getLooper().quit();
    }
    
    private class ServiceHandler extends Handler {
        ServiceHandler(Looper looper) {
            super(looper);
        }

        @Override
        public void handleMessage(Message msg) {
            Intent intent = (Intent) msg.obj;
            if (intent == null) return;
    
            SmsHandler smsHandler = SmsHandler.createFromIntent(
                SmsReceiveService.this, intent);
            // 处理短信
            smsHandler.analyseAndHandle();
            
            stopSelfResult(msg.arg1);
        }

    }
}

File: app/src/main/java/chenmc/sms/transaction/service/SmsReceiveService.kt
package chenmc.sms.transaction.service

import android.app.Service
import android.content.Intent
import android.os.*
import chenmc.sms.transaction.SmsExtractor
import chenmc.sms.transaction.SmsHandlerExecutor

/**
 * @author 明 明
 * Created on 2017-4-28.
 */

class SmsReceiveService : Service() {

    private lateinit var serviceHandler: Handler

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onCreate() {
        super.onCreate()
        // 新开一个线程进行处理
        val handlerThread = HandlerThread(
            javaClass.simpleName,
            Process.THREAD_PRIORITY_MORE_FAVORABLE
        )
        handlerThread.start()

        serviceHandler = ServiceHandler(handlerThread.looper)
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        val msg = serviceHandler.obtainMessage()
        // 保存 startId
        msg.arg1 = startId
        // 保存 intent 到 msg.obj
        msg.obj = intent
        serviceHandler.sendMessage(msg)

        return Service.START_NOT_STICKY
    }

    override fun onDestroy() {
        super.onDestroy()

        // Service 停止时停止线程
        serviceHandler.looper.quit()
    }

    private inner class ServiceHandler internal constructor(looper: Looper) : Handler(looper) {

        private val smsHandlerExecutor: SmsHandlerExecutor = SmsHandlerExecutor(this@SmsReceiveService)

        override fun handleMessage(msg: Message) {
            val intent = msg.obj as Intent

            // 从 Intent 中获取短信内容
            val sms = SmsExtractor.extractFromIntent(intent)
            // 短信执行器执行
            smsHandlerExecutor.execute(sms)

            // 使用 startId 停止 Service
            stopSelfResult(msg.arg1)
        }
    }
}
